geopandas/geodataframe.py

```diff
@@ -1422,6 +1408,35 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
 
     @property
     def _constructor(self):
+        geometry_default = getattr(self, "_geometry_column_name")
+        crs_default = getattr(self, "crs")
+
+        def _geodataframe_constructor_with_fallback(
+            data=None, index=None, crs=crs_default, geometry=geometry_default, **kwargs
+        ):
+            df = pd.DataFrame(data, index=index)
+            if geometry in df.columns:
+                geo_col = df[geometry]
+                if isinstance(geo_col, pd.DataFrame) or isinstance(
+                    geo_col.dtype, GeometryDtype
+                ):
+                    # DataFrame check is for when multiple columns are named
+                    # `geometry`. This case should error, but is passed through so that
+                    # we can have a specific error message for specific contexts,
+                    # e.g. concat, caught in __finalize__.
+
+                    df = GeoDataFrame(
+                        data=data, index=index, crs=crs, geometry=None, **kwargs
```

Comment:
We could also pass the already converted `df` here, instead of `data` + `index` + `kwargs` ?

Reply:
I thought there was a good reason for this, seems not, so I've converted this over.